src/target/riscv/riscv-013.c

```diff
@@ -3396,311 +3396,556 @@ static int write_memory_abstract(struct target *target, target_addr_t address,
+		enum gdb_regno b, int offset, unsigned int size, bool mprven)
+{
+	if (mprven && riscv_program_csrrsi(p, GDB_REGNO_ZERO, CSR_DCSR_MPRVEN,
+				GDB_REGNO_DCSR) != ERROR_OK)
+		return ERROR_FAIL;
+
+	if (riscv_program_load(p, d, b, offset, size) != ERROR_OK)
+		return ERROR_FAIL;
+
+	if (mprven && riscv_program_csrrci(p, GDB_REGNO_ZERO, CSR_DCSR_MPRVEN,
+				GDB_REGNO_DCSR) != ERROR_OK)
+		return ERROR_FAIL;
+
+	return ERROR_OK;
+}
+
+struct memory_access_info {
+	uint8_t *buffer_address;
+	target_addr_t target_address;
+	uint32_t element_size;
+	uint32_t increment;
+};
+
+static int read_memory_progbuf_inner_fill_progbuf(struct target *target,
+		uint32_t increment, uint32_t size, bool mprven);
+
+static int read_memory_progbuf_inner_startup(struct target *target,
+		target_addr_t address, uint32_t increment, uint32_t index);
+
+static int read_memory_progbuf_inner_loop_iteration(struct target *target,
+		struct memory_access_info access, uint32_t *index, uint32_t loop_count);
+
+static int read_word_from_dmi(struct target *target,
+		struct memory_access_info access, uint32_t index);
+
+static int read_word_from_s1(struct target *target,
+		struct memory_access_info access, uint32_t index);
 /**
  * Read the requested memory, taking care to execute every read exactly once,
  * even if cmderr=busy is encountered.
  */
-static int read_memory_progbuf_inner(struct target *target, target_addr_t address,
-		uint32_t size, uint32_t count, uint8_t *buffer, uint32_t increment)
+static int read_memory_progbuf_inner(struct target *target,
+		struct memory_access_info access, uint32_t count, bool mprven)
 {
-	RISCV013_INFO(info);
+	assert(count > 1 && "If count == 1, read_memory_progbuf_inner_one must be called");
 
-	int result = ERROR_OK;
+	if (read_memory_progbuf_inner_fill_progbuf(target, access.increment,
+				access.element_size, mprven) != ERROR_OK)
+		return ERROR_FAIL;
 
-	/* Write address to S0. */
-	result = register_write_direct(target, GDB_REGNO_S0, address);
-	if (result != ERROR_OK)
-		return result;
+	if (read_memory_progbuf_inner_startup(target, access.target_address,
+				access.increment, /*index*/ 0)
+			!= ERROR_OK)
+		return ERROR_FAIL;
+	/* The program in program buffer is executed twice during
+	 * read_memory_progbuf_inner_startup().
+	 * Here:
+	 * dmi_data[0:1] == M[address]
+	 * s1 == M[address + increment]
+	 * s0 == address + increment * 2
+	 * `count - 2` program executions are performed in this loop.
+	 * No need to execute the program any more, since S1 will already contain
+	 * M[address + increment * (count - 1)] and we can read it directly.
+	 */
+	const uint32_t loop_count = count - 2;
+
+	for (uint32_t index = 0; index < loop_count;)
+		if (read_memory_progbuf_inner_loop_iteration(target, access, &index,
+					loop_count) != ERROR_OK) {
+			dmi_write(target, DM_ABSTRACTAUTO, 0);
+			return ERROR_FAIL;
+		}
+	if (dmi_write(target, DM_ABSTRACTAUTO, 0) != ERROR_OK)
+		return ERROR_FAIL;
```

Comment:
Here you clear DM_ABSTRACTAUTO on one failure but not another. It might be simpler to clear it at the end of read_memory_progbuf() instead of going everywhere through the code where it might be necessary to clear it.

Reply:
I think I misread this.